package/com.unity.formats.usd/Runtime/Scripts/IO/SanitizedSamples.cs

```diff
@@ -0,0 +1,473 @@
+            {
+                var value = primvar.value as Vector2[];
+                Flatten(ref value, primvar.indices);
+                primvar.SetValue(value);
+                return;
+            }
+
+            if (primvar.value.GetType() == typeof(Vector3[]))
+            {
+                var value = primvar.value as Vector3[];
+                Flatten(ref value, primvar.indices);
+                primvar.SetValue(value);
+                return;
+            }
+
+            if (primvar.value.GetType() == typeof(Vector4[]))
+            {
+                var value = primvar.value as Vector4[];
+                Flatten(ref value, primvar.indices);
+                primvar.SetValue(value);
+                return;
+            }
+        }
+
+        internal static void TriangulateAttributes<T>(ref T[] values, int[] faceVertexCount, bool changeHandedness)
+        {
+            var newValues = new List<T>();
+            var last = 0;
+            for (var i = 0; i < faceVertexCount.Length; i++)
+            {
+                var next = last + 1;
+                for (var t = 0; t < faceVertexCount[i] - 2; t++)
+                    if (changeHandedness)
+                    {
+                        newValues.Add(values[next++]);
+                        newValues.Add(values[last]);
+                        newValues.Add(values[next]);
+                    }
+                    else
+                    {
+                        newValues.Add(values[last]);
+                        newValues.Add(values[next++]);
+                        newValues.Add(values[next]);
+                    }
+
+                last += faceVertexCount[i];
+            }
+
+            values = newValues.ToArray();
+        }
+
+        void ConvertInterpolationToFaceVarying<T>(ref T[] values, int[] vertexIndices, bool changeHandedness = false, TfToken interpolation = null)
+        {
+            if (values == null)
+                return;
+
+            if (interpolation == null)
+                interpolation = GuessInterpolation(values.Length);
+
+            if (interpolation == UsdGeomTokens.constant)
+            {
+                // Ignore, constant values are supported by the importer
+            }
+            if (interpolation == UsdGeomTokens.uniform) // 1 value per face
+            {
+                UniformToFaceVarying(ref values, vertexIndices.Length);
+            }
+            else if (interpolation == UsdGeomTokens.vertex || interpolation == UsdGeomTokens.varying) // 1 value per point
+            {
+                Flatten(ref values, faceVertexIndices);
+            }
+            else if (interpolation == UsdGeomTokens.faceVarying) // 1 value per vertex per triangle
+            {
+                TriangulateAttributes(ref values, originalFaceVertexCounts, changeHandedness);
+            }
+        }
+
+        /// <summary>
+        /// Returns the interpolation of an data array based on the number of elements
```

Comment:
It's worth noting that this method, in general, will not always be correct.

It can be correct for constant and uniform interpolation, but varying/vertex/facevarying will sometimes be ambiguous.

Reply:
You're right, indexed arrays could easily mess it up and they are probably frequentely used.  
Indexed Vertex/Varying arrays will be shorter that points.length and indexed FaceVarying arrays will be shorter than faceVertexIndices.length or even equal to points.length in a worst case scenario!
All those situation will cause the MeshImporter to fail creating a valid mesh.

Reply:
@jcowles Now the issue to solve is that we don't deserialize the interpolation type for normals and tangents (we don't do the indices either). Although we do it for the colors and uvs by defining them as Primvars. Normals and tangents are not technically USD primvars but do you see an issue if I set them as Primvars?

Reply:
Rereading the doc I had some of that wrong:
- normals can be "builtin" or primvars
- only primvar normals can be indexed
- builtin normals are subject to interpolation but not indexing

Primvar normals should normally take precedence on builtin normals but we dont support them at the moment. It leaves us with non indexed builtin normals for which the GuessInterpolation method should work.

@jcowles  What is the ambiguous situation you had in mind?

Reply:
Indexed primvars for sure, but there are degenerate non-indexed cases, where you have four faces and all faces share the same verts - uniform, vertex/varying, face varying are not distinguishable. Or more common cases, e.g. if you have a single face, then vertex & face-varying are not distinguishable.

I think what you have is fine, but it's worth calling this out.

Other primitives (such as curves) have even more aliasing.

Reply:
Got it thanks for the update. Adding that as comment in the code.